### ISSUE TITLE
Some questions and kind request for the code about training the explicit value model

### ISSUE THREAD

**Issue by andyclsr** [chadlzx/ivg#1]
Hi, great thanks for your excellent work and opening-source the code.

I have a question regarding the training procedure for explicit value models described in Equation 12. Specifically, did you train the intermediate states of the value model to fit the predictions of the reward model? If so, I’m curious whether there are additional implementation details that might shed more light on this process as it seems like a very simplified implementation version of the original paper.

As the original paper, "Controlled Decoding from Language Models," does not provide public code either, I was wondering if you could share further insights into this aspect or potentially any related code.

Thanks in advance! : )

**Comment by andyclsr** [comment]
Another problem with the implementation of the explicit value model, is it more convenient to implement it with trl.AutoModelForCausalLMWithValueHead? Are there any special reasons or points I missed for the code modeling_gpt2.py and modeling_llama.py?

Hope for your early reply!
Best

**Comment by chadlzx** [comment]
Thank you for your interest in our work and for your thoughtful questions! I’ll address them one by one:

1. **Training the Intermediate States of the Value Model**:
   Yes, we indeed implemented the training process where the intermediate states of the value model are fitted to the predictions of the reward model. This is a necessary step for training the value model effectively. Below is a snippet of the code we used for training the value function using the CD-FUDGE algorithm. As you can see, we extended the rewards for the entire sequence to each token's reward:

   ```python
   def loss(
       self,
       input_ids: torch.LongTensor,  # size = (B, L)
       attention_mask: torch.BoolTensor,  # size = (B, L)
       labels: torch.LongTensor,  # size = (B, L)
       ref_rewards: torch.FloatTensor,  # size = (B,)
   ) -> dict[str, torch.Tensor]:
       
       batch_size = input_ids.size(0)

       output: ScoreModelOutput = self.model(
           input_ids=input_ids,  # size = (B, L)
           attention_mask=attention_mask,  # size = (B, L)
       )
       logits = output.logits 

       if self.args.loss_type == "mse":
           shift_logits = logits[:, :-1, :].contiguous()
           shift_labels = input_ids[:, 1:].contiguous()
           shift_logits = shift_logits.gather(dim=-1, index=shift_labels.unsqueeze(-1)).squeeze(-1)  # size = (B, L-1)
           losses = []
           for i in range(batch_size):
               ref_reward_label = ref_rewards[i]
               truncated_rewards = shift_logits[i, labels[i][1:] != -100]
               ref_reward_label = ref_reward_label.to(truncated_rewards.dtype)
               mseloss = F.mse_loss(truncated_rewards, ref_reward_label.repeat(truncated_rewards.size(0)), reduction='mean') 
               losses.append(mseloss)
           loss = torch.stack(losses).mean()
       else:
           raise NotImplementedError
       return {
           'loss': loss,  # size = ()
       }
   ```

2. **Implementation of the Explicit Value Model**:
   Regarding your question about using `trl.AutoModelForCausalLMWithValueHead`, we modified the `safe-rlhf` framework to support our algorithm and training process. We haven’t experimented with TRL, but we believe there might be more straightforward ways to implement this, as you suggested. As for the "special reasons or points" you mentioned, I might have forgot some details from the paper, so if you could clarify further, I’d be happy to provide more insights.

**Comment by andyclsr** [comment]
Great thanks for the clear, informative, and prompt explanations! I understand now and wish you a great day! : )
